compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/FirSourceUtils.kt

```diff
@@ -0,0 +1,84 @@
+import com.intellij.psi.tree.TokenSet
+import org.jetbrains.kotlin.fir.*
+
+fun FirSourceElement.getChild(type: IElementType, index: Int = 0, depth: Int = -1): FirSourceElement? {
+    return getChild(setOf(type), index, depth)
+}
+
+fun FirSourceElement.getChild(types: TokenSet, index: Int = 0, depth: Int = -1): FirSourceElement? {
+    return getChild(types.types.toSet(), index, depth)
+}
+
+fun FirSourceElement.getChild(types: Set<IElementType>, index: Int = 0, depth: Int = -1): FirSourceElement? {
+    return when (this) {
+        is FirPsiSourceElement<*> -> {
+            getChild(types, index, depth)
+        }
+        is FirLightSourceElement -> {
+            getChild(types, index, depth)
+        }
+        else -> null
+    }
+}
+
+private fun FirPsiSourceElement<*>.getChild(types: Set<IElementType>, index: Int, depth: Int): FirSourceElement? {
+    val visitor = PsiElementFinderByType(types, index, depth)
+    return visitor.find(psi)?.toFirPsiSourceElement()
+}
+
+private fun FirLightSourceElement.getChild(types: Set<IElementType>, index: Int, depth: Int): FirSourceElement? {
+    val visitor = LighterTreeElementFinderByType(treeStructure, types, index, depth)
+    return visitor.find(lighterASTNode)?.toFirLightSourceElement(treeStructure)
+}
+
+fun FirSourceElement.getParent(type: IElementType, includingSelf: Boolean = false): FirSourceElement? {
+    return getParent(setOf(type), includingSelf)
+}
+
+fun FirSourceElement.getParent(types: TokenSet, includingSelf: Boolean = false): FirSourceElement? {
+    return getParent(types.types.toSet(), includingSelf)
+}
+
+fun FirSourceElement.getParent(types: Set<IElementType>, includingSelf: Boolean = false): FirSourceElement? {
+    return when (this) {
+        is FirPsiSourceElement<*> -> {
+            getParent(types, includingSelf)
+        }
+        is FirLightSourceElement -> {
+            getParent(types, includingSelf)
+        }
+        else -> null
+    }
+}
+
+private fun FirPsiSourceElement<*>.getParent(types: Set<IElementType>, includingSelf: Boolean): FirSourceElement? {
+    var parent: PsiElement? = if (includingSelf) psi else psi.parent
+    while (parent != null && parent.node.elementType !in types) {
+        parent = parent.parent
+    }
+    return parent?.toFirPsiSourceElement()
+}
+
+private fun FirLightSourceElement.getParent(types: Set<IElementType>, includingSelf: Boolean): FirSourceElement? {
+    var parent: LighterASTNode? = if (includingSelf) lighterASTNode else treeStructure.getParent(lighterASTNode)
+    while (parent != null && parent.tokenType !in types) {
+        parent = treeStructure.getParent(parent)
+    }
+    val startDelta = this.startOffset - lighterASTNode.startOffset
+    val endDelta = this.endOffset - lighterASTNode.endOffset
+    return parent?.toFirLightSourceElement(
+        treeStructure,
+        startOffset = startDelta + parent.startOffset,
+        endOffset = endDelta + parent.endOffset
+    )
```

Comment:
This looks like added/edited after I reviewed/approved the PR... (though I know the context: https://github.com/JetBrains/kotlin/pull/4199/commits/08556d5b8bba4cf3d0bbb8f751dd3f5a0d3bb147#diff-95541e526b759e380b19f0ebb0de115dc1a83296be0a0468f53ec9d48b9fda60)

It's better to make these a util somehow, and depending on which one goes first, you can update the remaining one to use that util?

Reply:
OK, added a util in this PR. If this PR is merged first. Then I will update the other.